cirq/ops/three_qubit_gates.py

```diff
@@ -145,6 +144,125 @@ def __str__(self) -> str:
         return 'CCZ**{}'.format(self._exponent)
 
 
+class ThreeQubitDiagonalGate(gate_features.ThreeQubitGate):
+    """A gate given by a diagonal 8x8 matrix."""
+
+    def __init__(self,
+                 a: float = 0.0,
+                 b: float = 0.0,
+                 c: float = 0.0,
+                 d: float = 0.0,
+                 e: float = 0.0,
+                 f: float = 0.0,
+                 g: float = 0.0,
+                 h: float = 0.0) -> None:
+        self._diag_angles = []  # type: List[float]
+        self._diag_angles.append(a)
+        self._diag_angles.append(b)
+        self._diag_angles.append(c)
+        self._diag_angles.append(d)
+        self._diag_angles.append(e)
+        self._diag_angles.append(f)
+        self._diag_angles.append(g)
+        self._diag_angles.append(h)
+        self._matrix = np.diag(
```

Comment:
Don't store the matrix, store the diagonal.

Reply:
Done.